### DISCUSSION TITLE
Question on experiments

### DISCUSSION THREAD

**Discussion by Aya-Badawy89** [gama-platform/gama#93] [Q&A] [answered]
Hello GAMA team,
Do we have a possibility in GAMA to conduct a **Monte Carlo simulation**? 
The agents in my model do some actions randomly during the simulation; then, they reach an outcome. 
I want to run several simulations all with the same parameters but explore how the outcome changes as a result of the stochasticity of variables.
To be more specific: I need multiple simulations to run sequentially and plot their results on the same chart or take their average. Is it possible?
I tried doing it by coupling the model into a macro-model which can manage the simulations and plot their results...etc. But, is there an easier solution?

**Comment by ptaillandier** [comment]
Hi,

Here attached an example model of what can be done (not sure it answers
your question): it runs 50 times the simulations (in batch mode) then
displays all the simulation results (for each step) and the mean value for
each step.

Cheers,

Patrick



Le jeu. 9 déc. 2021 à 07:38, Aya-Badawy89 ***@***.***> a
écrit :

**Reply by Aya-Badawy89** [reply]
Hi, Patrick,
I can't find the attachmed model in your reply.

**Comment by ptaillandier** [comment, marked as answer]
Hi,

Here the model:

```
model MonteCarlowithbatch

global {
float mean_val <- 10.0;
float standard_deviation_val <- 2.0;
float result <- gauss(mean_val,standard_deviation_val) ;
bool end <- false;

list<float> results;

reflex add_result {
result <- result + result/10.0 + (1.0 - rnd(1.0));
results << result;
end <- cycle = 10;
}
}


experiment explore_model_stochasticity type: batch until: end repeat: 50 {

list<list<float>> results_vals;

float marge <- 5.0;

reflex end_replication {

ask simulations {
myself.results_vals << self.results;
}
do update_outputs(true);
}

geometry median_lines(float min_v, float max_v, int num_c) {
list<point> pts;

loop i from: 0 to: num_c - 1 {
float v <-  median(results_vals collect each[i]);
pts << {marge  + i * (100 - marge * 2)/ num_c   , (100 - marge) - ((100 -
 marge*2) *  (v - min_v) / (max_v - min_v))};

i <- i + 1;
}
return line(pts);
}





list<geometry> axis {
list<geometry> geoms;
geoms << line([{marge, 100 - marge},{100 - marge, 100 - marge}]);
geoms << line([{marge, 100 - marge},{marge, marge}]) ;
return geoms;
}


list<geometry> result_lines(float min_v, float max_v) {
list<geometry> geoms;
loop c over: results_vals{
list<point> pts;
int i <- 0;
loop v over: c {
pts << {marge  + i * (100 - marge * 2)/ length(c)   , (100 - marge) - ((100
-  marge*2) *  (v - min_v) / (max_v - min_v))};
i <- i + 1;
}
geoms << line(pts) ;
}
return geoms;
}

permanent {
display all_simulation  {
graphics "all_simulations" {
if (not empty(results_vals)) {
int max_t <- length(results_vals[0]) - 1;
loop ax over: axis() {
draw ax  end_arrow: 1 color: #black;
}
draw "Mean results" rotate: -90 at: {4,50} anchor: #center font:
font("Helvetica", 18, #bold) color: #black;
draw "Time"  at: {50,98} anchor: #center font: font("Helvetica", 18, #bold)
color: #black;
draw "0"  at: {marge,96} anchor: #center font: font("Helvetica", 18, #bold)
color: #black;
draw "0"  rotate: -90  at: {marge,93} anchor: #center font:
font("Helvetica", 18, #bold) color: #black;

float max_v <- max(results_vals accumulate each);
float min_v <- min(results_vals accumulate each);
list<geometry> geoms <- result_lines(min_v,max_v);
loop i from: 0  to: length(geoms) - 1 {
draw geoms[i] color: #gray;
}
draw median_lines(min_v,max_v,max_t) + 0.2 color: #red ;

draw ""+ round(max_t)    at: {100 - marge,96} anchor: #center font:
font("Helvetica", 18, #bold) color: #black;
draw "" + max_v  with_precision 1 rotate: -90  at: {marge, marge} anchor:
#center font: font("Helvetica", 18, #bold) color: #black;



}
}
}
}

}
```

**Comment by sriramab** [comment]
Working on a video for this, just lazy to edit it :D ...

**Reply by RoiArthurB** [reply]
Please don't hesitate to post the video's link here once you publish it 🤗

**Comment by Aya-Badawy89** [comment]
Thank you very much. I have one more question, please.
In the batch experiments, simulations run in parallel. Is it possible to make them run one after the other. 
The problem is that I have **user_input dialogs** which display during simulation and I enter some variables. I don't want to enter the values 50 times. I want to insert them in the first run, and let the others take the same values. I don't know how I can do that!

**Reply by Aya-Badawy89** [reply]
As a response to my above question, I figured out a possible way by which the **user_input_dialogs** only appear in the first simulation while other simulations take the same values. Here is a dummy model for how we can do it:

```
model dummyModel

global {
	init {create girl;}	
}

species girl  {
	int var;
	reflex when: time=1 {
			if int(host) = 0 {
				map insert <- user_input_dialog ('Parameters', [enter('Value', 4)]);
				var <- int(insert['Value']);
			}
			else {
				loop while: host.peers[0].girl[0].var = 0 {}
				var <- host.peers[0].girl[0].var;		
			}
			write 'Sim num= ' + (int(host)) + ' and variable = ' + var; 		
	}
}

experiment dummy type: batch until: time=10 repeat:2  {
	
}
```

**Comment by ptaillandier** [comment]
Hi,

You can use the user_input_dialog operator in the init of the batch
experiment for that.

Cheers,

Patrick

Le dim. 12 déc. 2021 à 06:58, Aya-Badawy89 ***@***.***> a
écrit :

**Reply by Aya-Badawy89** [reply]
I have user_input_dialogs for each species at different points during the simulation. 
That's why I tried the above code so that whenever the simulations reach the user_input_operator, only the first simulation displays it. 
I'm not not sure if putting the dialogs in the init of the experiment will do the same because how will I relate them to the different situations during the simulation.
I am still working on it anyway. Let's see.
Thank you very much.

**Comment by Aya-Badawy89** [comment]
**Just to update what I did at the end for anyone who comes across this thread and has a similar challenge:**

My case in brief: I have a comodel where its agents run another model to compute a value during simulation. The model also displays user-input dialogs at different points while running. The challenge was how to run several simulations for this model and take the average. I tried 3 methods:
1) First method: [Create a macro-model and import the comodel in it](https://gama-platform.org/wiki/Comodel). Create one simulation of the comodel and run it, so that users can insert values of the user-input dialogs. Then, create the rest of simulations using the same values inserted by users in the first simulation so that users do not need to reply to user-input dialogs each time. 
This method threw an error. It is generally complicated because importing the comodel into a macromodel means that we are having three nested models which is not good for code interpretability.
2) Second method: [Use batch experiment](https://gama-platform.org/wiki/BatchExperiments). 
This method gave errors when applying it to my specific case.  Batch experiments run simulations simultaneously. The question was how to make input dialogs only appear in the first simulation while the rest take the same values. I tried the code in the previous comment, but it threw errors because simulations are not running at the same pace due to stochasticity which causes some changes in behaviour of agents. Also, simulations running at the same time may be confusing in a comodel.
3)  Third method: Use [create simulation](https://gama-platform.org/wiki/RunSeveralSimulations). 
This method worked. It helped me create simulations sequentially, display the user-dialogs only in the first, save the results of each simulation, then take the average at the end. Also, running the comodel seemed organized. The only draw-back is that I needed to write a lot of code for saving results and creating average...etc.

Those are my attempts and observations in case anyone finds them useful.
